cpp/arcticdb/version/version_core.cpp

```diff
@@ -1000,4 +963,109 @@ VersionedItem compact_incomplete_impl(
     return vit;
 }
 
+std::tuple<std::shared_ptr<PipelineContext>, ReadQuery, size_t, std::optional<size_t>> get_pre_defragmentation_info(
+        const std::shared_ptr<Store>& store,
+        const StreamId& stream_id,
+        const UpdateInfo& update_info,
+        const WriteOptions& options,
+        size_t segment_size) {
+    util::check(update_info.previous_index_key_.has_value(), "No latest undeleted version found for data compaction");
+
+    auto pipeline_context = std::make_shared<PipelineContext>();
+    pipeline_context->stream_id_ = stream_id;
+    pipeline_context->version_id_ = update_info.next_version_id_;
+
+    ReadQuery read_query;
+    read_indexed_keys_to_pipeline(store, pipeline_context, *(update_info.previous_index_key_), read_query, defragmentation_read_options_generator(options));
```

Comment:
Calling this low-level stuff scares me. Please write a test to double-check that check recursive normaliser works with this method?

Reply:
I am not sure what recursive normaliser is. Care to point me to a test method to let me refer to it?

Reply:
Do you mean check that recursively-normalized data can be compacted?

Reply:
If so, it is covered in the current test case.
I am also writing a new test case to address Qi's comment, which will utilize hypothesis. It should have covered this scenario in a better way.